af/detector/detector/detector.py

```diff
@@ -861,65 +212,316 @@ def load_country_name_map():
+# `blocking_status`.
+# Detection is performed on test_name+probe_cc+probe_asn+input aka "TCAI"
+# following query. It's easier to read the query from the bottom up:
+#  - Select data from the fastpath table for the last time window, counting
+#    the percentages of confirmed, anomaly-but-not-confirmed etc. --> As "new"
+#  - Select stored related moving averages etc. from blocking_status.
+#    FINAL is required to avoid duplicates due to table updates.
+#  - Join the two queries on TCAI
+#  - Run SELECT to compute new values:
+#    - The empty(blocking_status... parts are simply picking the right
+#      value in case the entry is missing from blocking_status (TCAI never
+#      seen before) or not in fastpath (no recent msmts for a given TCAI)
+#    - Weighted percentages e.g.
+#      (new_confirmed_perc * msmt count * μ + stored_perc * blocking_status.cnt * 𝜏) / totcnt
+#      Where: 𝜏 < 1, μ = 1 - t  are used to combine old vs new values,
+#       blocking_status.cnt is an averaged count of seen msmts representing how
+#       "meaningful" the confirmed percentage in the blocking_status table is.
+#    - Stability: compares the current and stored accessible/ confirmed/etc
+#        percentage. 1 if there is no change, 0 for maximally fast change.
+#        (Initialized as 0 for new TCAI)
+#  - Then run another SELECT to:
+#    - Set `status` based on accessible thresholds and stability
+#      We don't want to trigger a spurius change if accessibility rates
+#      floating a lot.
+#    - Set `change` if the detected status is different from the past
+#  - INSERT: Finally store the TCAI, the moving averages, msmt count, change,
+#    stability in blocking_status to use it in the next cycle.
+#
+#  As the INSERT does not returns the rows, we select them in extract_changes
+#  (with FINAL) to record them into `blocking_events`.
+#  If there are changes we then extract the whole history for each changed
+#  TCAI in rebuild_feeds.
+#
+@metrics.timer("run_detection")
+def run_detection(start_date, end_date, services) -> None:
+    if not conf.reprocess:
+        log.info(f"Running detection from {start_date} to {end_date}")
+
+    if 0 and conf.reprocess:
+        # Used for debugging
+        sql = """SELECT count() AS cnt FROM fastpath
+    WHERE test_name IN ['web_connectivity']
+    AND msm_failure = 'f'
+    AND measurement_start_time >= %(start_date)s
+    AND measurement_start_time < %(end_date)s
+    AND probe_asn = 135300 and probe_cc = 'MM' AND input = 'https://twitter.com/'
+        """
+        d = dict(start_date=start_date, end_date=end_date)
+        log_example = bool(query(sql, d)[0][0])
+
+    # FIXME:   new.cnt * %(mu)f + blocking_status.cnt * %(tau)f AS totcnt
+    sql = """
+INSERT INTO blocking_status (test_name, input, probe_cc, probe_asn,
+  confirmed_perc, pure_anomaly_perc, accessible_perc, cnt, status, old_status, change, stability)
+SELECT test_name, input, probe_cc, probe_asn,
```

Comment:
You can improve this query by dropping one level of nesting and moving `totcnt` into a CTE, like this:
```sql
WITH new.cnt + blocking_status.cnt AS totcnt
SELECT
 empty(blocking_status.test_name) ? new.test_name : blocking_status.test_name AS test_name,
 empty(blocking_status.input) ? new.input : blocking_status.input AS input,
 empty(blocking_status.probe_cc) ? new.probe_cc : blocking_status.probe_cc AS probe_cc,
 (blocking_status.probe_asn = 0) ? new.probe_asn : blocking_status.probe_asn AS probe_asn,
 new.confirmed_perc    * new.cnt/totcnt * 0.015 + blocking_status.confirmed_perc    * blocking_status.cnt/totcnt * 0.985 AS confirmed_perc,
 new.pure_anomaly_perc * new.cnt/totcnt * 0.015 + blocking_status.pure_anomaly_perc * blocking_status.cnt/totcnt * 0.985 AS pure_anomaly_perc,
 new.accessible_perc   * new.cnt/totcnt * 0.015 + blocking_status.accessible_perc   * blocking_status.cnt/totcnt * 0.985 AS accessible_perc,
 new.cnt * 0.015 + blocking_status.cnt * 0.985 AS cnt,
  multiIf(
    accessible_perc < 80 AND stability > 0.95, 'BLOCKED',
    accessible_perc > 95 AND stability > 0.97, 'OK',
    blocking_status.status) AS status,
  blocking_status.status AS old_status,
  if(status = old_status, change * 0.985, 1) AS change,
 if(new.cnt > 0,
  cos(3.14/2*(new.accessible_perc - blocking_status.accessible_perc)/100) * 0.7 + blocking_status.stability * 0.3,
  blocking_status.stability) AS stability
FROM blocking_status FINAL
FULL OUTER JOIN
(
 SELECT test_name, input, probe_cc, probe_asn,
  countIf(confirmed = 't') * 100 / cnt AS confirmed_perc,
  countIf(anomaly = 't') * 100 / cnt - confirmed_perc AS pure_anomaly_perc,
  countIf(anomaly = 'f') * 100 / cnt AS accessible_perc,
  count() AS cnt,
  0 AS stability
 FROM fastpath
 WHERE test_name IN ['web_connectivity']
 AND msm_failure = 'f'
 AND measurement_start_time >= %(start_date)s
 AND measurement_start_time < %(end_date)s
 AND input IN %(urls)s
 GROUP BY test_name, input, probe_cc, probe_asn
) AS new
ON
 new.input = blocking_status.input
 AND new.probe_cc = blocking_status.probe_cc
 AND new.probe_asn = blocking_status.probe_asn
 AND new.test_name = blocking_status.test_name
```